### ISSUE TITLE
With a -T switch: untie attempted while 1 inner references still exist

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10920]
Migrated from [rt.perl.org#81230](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=81230) (status was 'resolved')

Searchable as RT81230$

**Comment by p5pRT** [comment]
### From Mark.Martinec@ijs.si

#### Created by Mark.Martinec@ijs.si



Switching from perl 5\.12\.2 to 5\.13\.8 I noticed SpamAssassin's
self\-test producing warnings\. A test case distills down to&#8203;:

$ perl \-Twe 'use DB\_File; my %db;
&nbsp;  tie\(%db\,"DB\_File"\,"0\.dat"\) or die $\!;
&nbsp;  $db\{"k"\}="x"; untie %db if $db\{"k"\}'

produces&#8203;:

&nbsp;  untie attempted while 1 inner references still exist at \-e line 1\.

Leaving out the \-T switch avoids the problem\!

I'm not sure whether this is an issue with a module DB\_File
or with core tie handling\. I don't see an error in usage\.
Everything else seems to be running fine\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.13.8:

Configured by mark at Thu Dec 23 19:15:09 CET 2010.

Summary of my perl5 (revision 5 version 13 subversion 8) configuration:
   
  Platform:
    osname=freebsd, osvers=7.2-release-p2, archname=amd64-freebsd
    uname='freebsd dorothy.ijs.si 7.2-release-p2 freebsd 7.2-release-p2 #0: wed jul 15 15:45:26 cest 2009 lesi@dorothy.ijs.si:usrobjusrsrcsysdorothy amd64 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include',
    optimize='-O',
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.2.1 20070719  [FreeBSD]', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E  -fstack-protector -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lgdbm -lm -lcrypt -lutil -lc
    perllibs=-lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.13.8:
    /usr/local/lib/perl5/site_perl/5.13.8/amd64-freebsd
    /usr/local/lib/perl5/site_perl/5.13.8
    /usr/local/lib/perl5/5.13.8/amd64-freebsd
    /usr/local/lib/perl5/5.13.8
    /usr/local/lib/perl5/site_perl
    .


Environment for perl 5.13.8:
    HOME=/root
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Dec 23\, 2010 at 06&#8203;:23&#8203;:54PM \-0800\, Mark Martinec wrote&#8203;:

Thanks for the report\.

> Switching from perl 5\.12\.2 to 5\.13\.8 I noticed SpamAssassin's
> self\-test producing warnings\. A test case distills down to&#8203;:
> 
> $ perl \-Twe 'use DB\_File; my %db;
>   tie\(%db\,"DB\_File"\,"0\.dat"\) or die $\!;
>   $db\{"k"\}="x"; untie %db if $db\{"k"\}'
> 
> produces&#8203;:
> 
>   untie attempted while 1 inner references still exist at \-e line 1\.
> 
> Leaving out the \-T switch avoids the problem\!
> 
> I'm not sure whether this is an issue with a module DB\_File
> or with core tie handling\. I don't see an error in usage\.
> Everything else seems to be running fine\.

I don't know either\. I ran git bisect with this&#8203;:

\#\!/bin/sh
git clean \-dxf
touch \.patchnum
touch \.sha1
touch unpushed\.h
\# If you can use ccache\, add \-Dcc=ccache\\ gcc \-Dld=gcc to the Configure line
\# if Encode is not needed for the test\, you can speed up the bisect by
\# excluding it from the runs with \-Dnoextensions=Encode
sh Configure \-des \-Dusedevel \-Uusethreads \-Doptimize="\-g" \-Dcc=ccache\\ gcc \-Dld=gcc \-Dnoextensions=IPC/SysV\\ Encode
test \-f config\.sh || exit 125
\# Correct makefile for newer GNU gcc
perl \-ni \-we 'print unless /\<\(?&#8203;:built\-in|command\)/' makefile x2p/makefile
\# if you just need miniperl\, replace test\_prep with miniperl
make \-j5 test\_prep
\[ \-x \./perl \] || exit 125
\./perl \-Ilib \-Twe '$SIG\{\_\_WARN\_\_\} = sub \{die @&#8203;\_\}; use DB\_File; my %db;  tie\(%db\,"DB\_File"\,"0\.dat"\) or die $\!; $db\{"k"\}="x"; untie %db if $db\{"k"\}'
ret=$?
\[ $ret \-gt 127 \] && ret=127
git clean \-dxf
exit $ret


and it identifies the behaviour change as being due to this commit&#8203;:

commit efaf36747029c85b4d8825318cb4d485a0bb350e
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Sun Apr 25 00&#8203;:56&#8203;:32 2010 \+0100

&nbsp;    add Perl\_magic\_methcall
&nbsp;    
&nbsp;    Add a new function that wraps the setup needed to call a magic method like
&nbsp;    FETCH \(the existing S\_magic\_methcall function has been renamed
&nbsp;    S\_magic\_methcall1\)\.
&nbsp;    
&nbsp;    There is one functional change\, done mainly to allow for a single clean
&nbsp;    wrapper function\, and that is that the method calls are no longer wrapped
&nbsp;    with SAVETMPS/FREETMPS\. Previously only about half of them had this\, so
&nbsp;    some relied on the caller to free\, some didn't\. At least we're consistent
&nbsp;    now\. Doing it this way is necessary because otherwise magic\_methcall\(\)
&nbsp;    can't return an SV \(eg for POP\) because it'll be a temp and get freed by
&nbsp;    FREETMPS before it gets returned\. So you'd have to copy everything\, which
&nbsp;    would slow things down\.



I tried substituting DB\_File in your example with Tie&#8203;::StdHash and there is
no warning\. So I'm wondering if it's something to do with the implementation
of DB\_File\. I'm afraid that I don't have any insight into this\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Dec 24\, 2010 at 05&#8203;:00&#8203;:30PM \+0000\, Nicholas Clark wrote&#8203;:
> On Thu\, Dec 23\, 2010 at 06&#8203;:23&#8203;:54PM \-0800\, Mark Martinec wrote&#8203;:
> 
> Thanks for the report\.
> 
> > Switching from perl 5\.12\.2 to 5\.13\.8 I noticed SpamAssassin's
> > self\-test producing warnings\. A test case distills down to&#8203;:
> > 
> > $ perl \-Twe 'use DB\_File; my %db;
> >   tie\(%db\,"DB\_File"\,"0\.dat"\) or die $\!;
> >   $db\{"k"\}="x"; untie %db if $db\{"k"\}'
> > 
> > produces&#8203;:
> > 
> >   untie attempted while 1 inner references still exist at \-e line 1\.
> > 
> > Leaving out the \-T switch avoids the problem\!
> > 
> > I'm not sure whether this is an issue with a module DB\_File
> > or with core tie handling\. I don't see an error in usage\.
> > Everything else seems to be running fine\.
> 
> I don't know either\. I ran git bisect with this&#8203;:
> 
> \#\!/bin/sh
> git clean \-dxf
> touch \.patchnum
> touch \.sha1
> touch unpushed\.h
> \# If you can use ccache\, add \-Dcc=ccache\\ gcc \-Dld=gcc to the Configure line
> \# if Encode is not needed for the test\, you can speed up the bisect by
> \# excluding it from the runs with \-Dnoextensions=Encode
> sh Configure \-des \-Dusedevel \-Uusethreads \-Doptimize="\-g" \-Dcc=ccache\\ gcc \-Dld=gcc \-Dnoextensions=IPC/SysV\\ Encode
> test \-f config\.sh || exit 125
> \# Correct makefile for newer GNU gcc
> perl \-ni \-we 'print unless /\<\(?&#8203;:built\-in|command\)/' makefile x2p/makefile
> \# if you just need miniperl\, replace test\_prep with miniperl
> make \-j5 test\_prep
> \[ \-x \./perl \] || exit 125
> \./perl \-Ilib \-Twe '$SIG\{\_\_WARN\_\_\} = sub \{die @&#8203;\_\}; use DB\_File; my %db;  tie\(%db\,"DB\_File"\,"0\.dat"\) or die $\!; $db\{"k"\}="x"; untie %db if $db\{"k"\}'
> ret=$?
> \[ $ret \-gt 127 \] && ret=127
> git clean \-dxf
> exit $ret
> 
> 
> and it identifies the behaviour change as being due to this commit&#8203;:
> 
> commit efaf36747029c85b4d8825318cb4d485a0bb350e
> Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> Date&#8203;:   Sun Apr 25 00&#8203;:56&#8203;:32 2010 \+0100
> 
>     add Perl\_magic\_methcall
>     
>     Add a new function that wraps the setup needed to call a magic method like
>     FETCH \(the existing S\_magic\_methcall function has been renamed
>     S\_magic\_methcall1\)\.
>     
>     There is one functional change\, done mainly to allow for a single clean
>     wrapper function\, and that is that the method calls are no longer wrapped
>     with SAVETMPS/FREETMPS\. Previously only about half of them had this\, so
>     some relied on the caller to free\, some didn't\. At least we're consistent
>     now\. Doing it this way is necessary because otherwise magic\_methcall\(\)
>     can't return an SV \(eg for POP\) because it'll be a temp and get freed by
>     FREETMPS before it gets returned\. So you'd have to copy everything\, which
>     would slow things down\.
> 
> 
> 
> I tried substituting DB\_File in your example with Tie&#8203;::StdHash and there is
> no warning\. So I'm wondering if it's something to do with the implementation
> of DB\_File\. I'm afraid that I don't have any insight into this\.

It's indirectly due to DB\_File tainting its tied object\. It can be
reproduced in plain perl by doing something similar\. Fixed by the
following commit&#8203;:

commit 8985fe98dcc5c0af2fadeac15dfbc13f553ee7fc
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Thu Dec 30 10&#8203;:32&#8203;:44 2010 \+0000
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Thu Dec 30 10&#8203;:51&#8203;:45 2010 \+0000

&nbsp;    Better handling of magic methods freeing the SV
&nbsp;    
&nbsp;    This is a fix for RT \#81230 \(and more\)\. Currently\, mg\_get\(\) works around
&nbsp;    the case where the called magic \(e\.g\. FETCH\) frees the magic SV\. It does
&nbsp;    this by unconditionally pushing the SV on the tmps stack before invoking
&nbsp;    the method\.
&nbsp;    
&nbsp;    There are two issues with this\. Firstly\, it may artificially extend the
&nbsp;    life of the SV\. This was the root of the problem with \#81230\. There\, the
&nbsp;    DB\_File code\, under \-T\, created a tainted tied object\. Accessing the
&nbsp;    object \(within FETCH as it happens\)\, caused mg\_get\(\) to be invoked on the
&nbsp;    object \(due to the taint magic\)\, and thus extend the life of the object\.
&nbsp;    This then caused c\<untie %h if $h\{k\}> to give the warning
&nbsp;        untie attempted while 1 inner references still exist\.
&nbsp;    This only became noticeable after efaf36747029c85b4d8825318cb4d485a0bb350e\,
&nbsp;    which stopped wrapping magic method calls in SAVETMPS/FREETMPS\.
&nbsp;    
&nbsp;    The second issue issue that this protection only applies to mg\_get\(\);
&nbsp;    functions like mg\_set\(\) can still segfault if the SV is deleted\.
&nbsp;    
&nbsp;    This commit fixes both problems as follows&#8203;:
&nbsp;    
&nbsp;    First\, the protection mechanism is moved out of mg\_get\(\) and into
&nbsp;    save\_magic\(\) / restore\_magic\(\)\, so that it protects more things\.
&nbsp;    Secondly\, the protection is now&#8203;:
&nbsp;    
&nbsp;    \* in save\_magic\(\)\, SvREFCNT\_inc\(\) the SV\, thus protecting it from being
&nbsp;      freed during FETCH \(or whatever\)
&nbsp;    
&nbsp;    \* in restore\_magic\(\)\, SvREFCNT\_dec\(\) the SV\, undoing the protection
&nbsp;      without extending the life of the SV\, \*except\* if the refcount is
&nbsp;      1 \(ie FETCH tried to free it\)\, then push it on the mortals stack
&nbsp;      to extend it life a bit so our callers wont choke on it\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    mg\.c
&nbsp;    M    t/op/taint\.t
&nbsp;    M    t/op/tie\.t


\-\- 
Red sky at night \- gerroff my land\!
Red sky at morning \- gerroff my land\!
&nbsp;    \-\- old farmers' sayings \#14

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
